File: DP/DP/12865-2024.swift
//
//  12865-2024.swift
//  DP
//
//  Created by sujeong on 2024/07/01.
//  Copyright © 2024 임수정. All rights reserved.
//

import Foundation

let inputs = readLine()?.split(separator: " ")
let n:Int = Int(inputs![0])!
let cap:Int = Int(inputs![1])!

var arr: [[Int]] = Array(repeating: Array(repeating: 0, count: cap+1), count: n+1)

var weight = 0
var value = 0

for i in 1...n {
    let input = readLine()?.split(separator: " ")
    weight = Int(input![0])!
    value = Int(input![1])!
    
    for j in 1...cap {
        if ( j < weight || arr[i-1][j] > arr[i-1][j-weight]+value ) {
            arr[i][j] = arr[i-1][j]
        } else {
            arr[i][j] = arr[i-1][j-weight]+value
        }
    }
}

print(arr[n][cap])

File: DP/DP/12865-2024.cpp
//
//  12865-2024.cpp
//  DP
//
//  Created by sujeong on 2024/07/01.
//  Copyright © 2024 임수정. All rights reserved.
//

#include <iostream>
#include <vector>
using namespace std;

int main(){
    
    int n, capacity;
    cin >> n >> capacity;
    
    vector<vector<int>> knapsack(n+1, vector<int>(capacity+1, 0));
    
    for(int item = 1 ; item <= n ; item++){
        int weight, value;
        cin >> weight >> value;
        
        for(int cur_cap = 1 ; cur_cap <= capacity ; cur_cap++){
            int total_value_without_item = knapsack[item-1][cur_cap];
            int total_value_with_item = knapsack[item-1][cur_cap - weight] + value;
            
            if(cur_cap < weight || total_value_with_item < total_value_without_item){
                // 배낭에 item 넣지 않음
                knapsack[item][cur_cap] = knapsack[item-1][cur_cap];
            }else{
                // 배낭에 item 넣음
                knapsack[item][cur_cap] = total_value_with_item;
            }
        }
    }
    
    cout << knapsack[n][capacity];
    
    return 0;
}
